cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringClient.java

```diff
@@ -94,14 +95,24 @@ public void createTethering(PeerInfo peerInfo) throws URISyntaxException, IOExce
    */
   public TetheringControlResponseV2 pollControlChannel(PeerInfo peerInfo,
                                                        TetheringControlChannelRequest channelRequest)
-    throws IOException, URISyntaxException {
+    throws IOException, URISyntaxException, NotFoundException {
     String content = GSON.toJson(channelRequest);
     String path = appendPath(Objects.requireNonNull(peerInfo.getEndpoint()),
-                                            CONNECT_CONTROL_CHANNEL + instanceName);
+                             CONNECT_CONTROL_CHANNEL + instanceName);
     URI peerUri  = new URI(peerInfo.getEndpoint()).resolve(path);
     HttpResponse resp = sendHttpRequest(remoteAuthenticator, HttpMethod.POST,
-                                                        peerUri, content);
-    if (resp.getResponseCode() != HttpURLConnection.HTTP_OK) {
+                                        peerUri, content);
+    if (resp.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+      String body = resp.getResponseBodyAsString(StandardCharsets.UTF_8);
+      TetheringControlResponseV2 controlResponse;
+      try {
+        controlResponse = GSON.fromJson(body, TetheringControlResponseV2.class);
+      } catch (Exception e) {
+        // 404 was not returned by the tethering server. Return the response body as the error message
+        throw new NotFoundException(body);
+      }
+      throw new NotFoundException(controlResponse);
```

Comment:
the object in the exception is normally expected to be the object that couldn't be found. In this particular case, I think it would be better to return a TetheringControlResponseV2 with not_found status instead of throwing an exception. This makes it simpler for the caller to handle things.

Reply:
fixed